### PR TITLE
Free old root environment, delete pools, change delevalstr argument type.

### PR DESCRIPTION
**Pull request by angularorbit**
The new `deltree` function header, requires `del` function to have a generic argument type. So I had to change the delevalstr argument type to a `void *`

### CODE REVIEW

File: env.c [hunk 1]
```diff
@@ -19,11 +19,19 @@ static struct treenode *pools;
...
+static void delrule(void *);
+
 void
 envinit(void)
 {
-	/* TODO: delete old root environment and pools (in case we rebuilt
-	 * build.ninja). for now, we leak memory. */
+	/* free old rootenv in case rebuilt the manifest */
+	if (rootenv) {
+		deltree(rootenv->bindings, free);
```
Comment (angularorbit) [inline env.c]: rootenv->bindings values are strings, `string struct` is freed using free.

File: env.c [hunk 2]
```diff
@@ -178,6 +186,19 @@ mkrule(char *name)
 	return r;
 }
 
+static void
+delrule(void *p)
+{
+	struct rule *r = p;
+	
+	if(!r)
+		return;
```
Comment (michaelforney) [inline env.c]: Is it ever the case that we have a NULL rule here? I'd prefer to remove this check if it is unneeded.

File: env.c [hunk 3]
```diff
@@ -19,11 +19,19 @@ static struct treenode *pools;
...
 {
-	/* TODO: delete old root environment and pools (in case we rebuilt
-	 * build.ninja). for now, we leak memory. */
+	/* free old rootenv in case rebuilt the manifest */
+	if (rootenv) {
+		deltree(rootenv->bindings, free);
+		deltree(rootenv->rules, delrule);
+		deltree(pools, delpool);
+		free(rootenv);
+	}
```
Comment (michaelforney) [inline env.c]: I think we still leak memory if `subninja` was used.

Fixing this would probably involve keeping a list of all environments, and traversing the list here. What do you think?

File: env.h [hunk 4]
```diff
@@ -16,21 +16,30 @@ struct pool {
...
 struct rule *mkrule(char *);
+/* add to a rule a variable and its value */
 void ruleaddvar(struct rule *, char *, struct evalstring *);
 
 /* create a new pool with the given name */
 struct pool *mkpool(char *);
 /* lookup a pool by name. fails if it does not exist */
 struct pool *poolget(char *);
 
+/* return an edge's variable evaluated, optionally shell-escaped */
```
Comment (michaelforney) [inline env.h]: These are all fine (thanks!), but could you move these comment changes to a separate commit?

File: util.c [hunk 5]
```diff
@@ -135,8 +135,9 @@ mkstr(size_t n)
 }
 
 void
-delevalstr(struct evalstring *str)
+delevalstr(void *pn)
```
Comment (michaelforney) [inline util.c]: I think we should come up with a consistent  name for this parameter. Perhaps `ptr` or `data`?

File: env.c [hunk 6]
```diff
@@ -19,11 +19,19 @@ static struct treenode *pools;
 
 static void addpool(struct pool *);
```
Comment (michaelforney) [inline env.c]: This blank line can be removed, just group all the forward declarations together.

File: env.c (hunk 2 above)
Comment (angularorbit) [reply env.c]: I haven't read the whole code base, so I don't for sure. probably unneeded, so I'll remove both this check and the rule-name check, since `mkrule` allocates a name.

File: util.c (hunk 5 above)
Comment (angularorbit) [reply util.c]: I'll use ptr

File: env.c (hunk 6 above)
Comment (angularorbit) [reply env.c]: Fair enough.

File: env.c (hunk 3 above)
Comment (angularorbit) [reply env.c]: I haven't read the whole codebase yet, so I don't know which environments I need to remove. Also no idea what `subninja` is. I'll write a TODO for the extra leaked memory, but keep the code as is.

Comment (michaelforney) [reply env.c]: The root environment that contains definitions in `build.ninja`. If `build.ninja` contains any [`subninja`](https://ninja-build.org/manual.html#ref_scope) lines, new environments will be created with the root environment as a parent, and so on.

Since we only have pointers back to the parent, there is currently no way to traverse the tree of environments. I can fix this in a follow-up commit. A TODO is fine for now.

File: env.c (hunk 2 above)
Comment (michaelforney) [reply env.c]: The rule name check should not be needed either.

I believe the same should be true for pools as well.

File: env.h (hunk 4 above)
Comment (angularorbit) [reply env.h]: Ok, done!

**Comment by angularorbit** [comment]
How's it looking?

File: env.c (hunk 2 above)
Comment (angularorbit) [reply env.c]: just edited the pools. was experimenting with git rebase, and it merged the amended commit smoothly.

Comment (angularorbit) [reply env.c]: Also after I had split the comment into 2 and pushed the first, I had lost my comments, but was able to find them with `git reflog`. That's next level!

Comment (michaelforney) [reply env.c]: Thanks. Yeah, `git rebase` and `git reflog` are very handy tools, indeed.

**Comment by michaelforney** [comment]
Looking great, thanks!

I was doing some testing with valgrind, and found a couple of issues:
- The `phonyrule` and `consolepool` should not be freed, since they are globals. I think we could just have `delrule` and `delpool` only free when it is not one of those.
- Since the rule and pool names are also used as the tree key, we end up freeing those twice. I think we can solve this by skipping the name free in `delrule` and `delpool`. If we did it the other way and remove the key free, then environment and rule variable names wouldn't get freed.

**Comment by angularorbit** [comment]
Rebased again. We need a taste suite. Also I need to start using a debugger.

**Comment by michaelforney** [comment]
Looks like there is still an issue, since string literals for "console" and "phony" are freed. Best thing I can think of is to pass separate functions for freeing the key and the value, so I pushed a commit to your branch that does that.

Does this seem okay to you?

**Comment by michaelforney** [comment]
Another possibility is to not actually insert the `phonyrule` and `consolepool` into the tree, just have a special case for them in `envrule` and `poolget`.

**Comment by angularorbit** [comment]
I think passing a tree key-free function makes sense, because `tree` is not responsible for allocating the keys, and shouldn't be responsible for freeing them.

Whether to remove `phonyrule` and `consolepool` from the tree, and special cases, I don't know. I'll leave the decision to you.

**Comment by michaelforney** [comment]
Whoops, I pushed the change before I read your edit. I don't understand what you mean about which parts of the tree need to be freed and which don't. It just depends on whether the pool/rule was a special static one.

Feel free to send another PR if you think we should use the other approach.
